toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -660,3 +723,95 @@ func getPartUUID(device string) (uuid string, err error) {
 	uuid = strings.TrimSpace(stdout)
 	return
 }
+
+func checkSectorSizeFile(sectorFile string) (sectorSize uint64, err error) {
+	if exists, ferr := file.PathExists(sectorFile); ferr != nil {
+		logger.Log.Errorf("Error accessing sector size file %s", sectorFile)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("could not find the hw sector size file %s to obtain the sector size of the system", sectorFile)
+		return
+	}
+
+	fileContent, err := file.ReadLines(sectorFile)
+	if err != nil {
+		logger.Log.Errorf("Failed to read from %s: %s", sectorFile, err)
+		return
+	}
+
+	// sector file should only have one line, return error if not
+	if len(fileContent) != 1 {
+		err = fmt.Errorf("%s has more than one line", sectorFile)
+		return
+	}
+
+	sectorSize, err = strconv.ParseUint(fileContent[0], 10, 64)
+	return
+}
+
+func getSectorSize(diskDevPath string) (logicalSectorSize, physicalSectorSize uint64, err error) {
+	// Grab the specific disk name from /dev/xxx
+	diskName := diskDevPath[5:len(diskDevPath)]
+
+	hw_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/hw_sector_size", diskName)
+	physical_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/physical_block_size", diskName)
+
+	logicalSectorSize, err = checkSectorSizeFile(hw_sector_size_file)
+	if err != nil {
+		return
+	}
+
+	physicalSectorSize, err = checkSectorSizeFile(physical_sector_size_file)
+	return
+}
+
+func alignSectorAddress(sectorAddr, logicalSectorSize, physicalSectorSize uint64) (alignedSector uint64) {
+	// Need to make sure that starting sector of a partition is aligned based on the physical sector size of the system.
+	// For example, suppose the physical sector size is 4096. If the input start sector is 40960001, then this is misaligned,
+	// and need to be elevated to the next aligned address, which is (40960001/4096 + 1)*4096 = 4100096.
+
+	// We do need to take care of a special case, which is the first partition (normally boot partition) might be less than
+	// the physical sector size. In this case, we need to check whether the start sector is a multiple of 1 MiB.
+	if sectorAddr < physicalSectorSize {
+		if sectorAddr%(oneMiBtoBytes/logicalSectorSize) == 0 {
+			alignedSector = sectorAddr
+		}
+	} else if (sectorAddr % physicalSectorSize) == 0 {
+		alignedSector = sectorAddr
+	} else {
+		alignedSector = (sectorAddr/physicalSectorSize + 1) * physicalSectorSize
+	}
+
+	return
+}
+
+func obtainPartitionDetail(partitionIndex int, hasExtendedPartition bool) (partType string, partitionNumber int) {
+	const (
+		primaryPartition              = "primary"
+		extendedPartition             = "extended"
+		logicalPartition              = "logical"
+		normalPartitionNumberToIndex  = 1
```

Comment:
>normalPartitionNumberToIndex [](http://example.com/codeflow?start=2&length=28)

I like named constants instead of raw numbers but isn't this the other way around? As in "index -> normal/logical partition number"? Or even `indexOffsetForNormalPartitionNumber`?

Reply:
Fixed